src/clj/clojurians_log/routes.clj

```diff
@@ -40,17 +53,30 @@
 
      ;; https://clojurians-log.clojureverse.org/clojure/2017-11-15.html
      (GET "/:channel/:date.html" [channel date :as request]
-       (let [db (d/db conn)]
-         (-> request
-             context
-             (assoc :data/channel (queries/channel db channel)
-                    :data/channels (queries/channel-list db date)
-                    :data/messages (queries/channel-day-messages db channel date)
-                    :data/channel-days (queries/channel-days db channel)
-                    :data/title (str channel " " date " | Clojurians Slack Log")
-                    :data/date date)
-             views/log-page
-             response/render)))
+       (let [config (var-get (resolve 'clojurians-log.application/config))
+             cache-time (get-in config [:message-page :cache-time])]
+         ;; Since we're displaying a log, presumably, all of the content is permanently cachable
+         ;; Message page content also most likely have not changed and does not require any processing/page-generation.
+         ;; In those cases, skip page generate entirely, instead of generating the contents, then have the ring
+         ;; middleware respond with a 304.
+         (if (and (not (zero? cache-time))
+                  (not (message-page-might-have-updated? date (get-in request [:headers "if-modified-since"]))))
+           (-> (ring.util.response/response nil)
+               (ring.util.response/status 304))
+
+           (let [db (d/db conn)]
+             (-> request
+                 context
+                 (assoc :data/channel (queries/channel db channel)
+                        :data/channels (queries/channel-list db date)
+                        :data/messages (queries/channel-day-messages db channel date)
+                        :data/channel-days (queries/channel-days db channel)
+                        :data/title (str channel " " date " | Clojurians Slack Log")
+                        :data/date date)
+                 views/log-page
+                 (assoc-in [:response/headers "Cache-Control"] (str "public, max-age: " cache-time))
```

Comment:
@plexus 

I think typical ring apps would generate an empty response map, transform the :status, :headers, and :body contents, the pass that back at the end of the handler.

The existing code, however, seems to want to assoc those data into the request map at :response/status, :response/headers, and :response/body, then extract those values into a response map in `response/render`. It's a bit strange to start mixing request data with response data this way. Is this structure just something that was inherited from old code? Or is there some special consideration behind this?

Reply:
That's a deliberate choice, partly inspired by how Pedestal does it. There's just a single `context` map that gets passed around. It's a bit of an experiment but I like it so far, but it's not set in stone, I'm open to considering other ways of doing things.

I'm also not married to Compojure, and would in fact prefer bidi + yada, but so far the app is small enough that Compojure will do.